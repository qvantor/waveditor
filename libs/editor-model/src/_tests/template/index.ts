export * as BasicCase from './basic-case';
export * as HostTemplate from './host-template';
