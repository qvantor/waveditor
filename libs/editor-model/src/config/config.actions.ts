import { getElementsFromElementFontRelationByFontId } from '../relations';
import { BuilderContext } from '../types';
import { ConfigFont } from './config.types';
import { createConfigFont } from './config.creator';

export const removeConfigFontById =
  ({ model: { relations, config }, module: { undoRedo } }: BuilderContext) =>
  (fontId: string) => {
    undoRedo.startBunch();
    const elementsWithFont = getElementsFromElementFontRelationByFontId(fontId)(
      relations.getValue()
    );
    elementsWithFont.forEach(relations.actions.removeElementFontRelation);
    config.actions.removeFont(fontId);
    undoRedo.endBunch();
  };

export const addConfigFont =
  ({ model: { relations, config }, module: { undoRedo } }: BuilderContext) =>
  (font: Omit<ConfigFont, 'id'>, element: string) => {
    const newFont = createConfigFont(font);
    undoRedo.startBunch();
    config.actions.addFont(newFont);
    relations.actions.addElementFontRelation({
      font: newFont.id,
      element,
    });
    undoRedo.endBunch();
  };
