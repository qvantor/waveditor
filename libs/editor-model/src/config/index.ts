export type {
  Config,
  ConfigFont,
  ConfigStoreUndoRedoEvent,
  FontChangedPayload,
} from './config.types';
export type { ConfigStore } from './config';
export { configStoreConstructor } from './config';
export { createInitialConfig } from './config.creator';
export * from './config.selectors';
export * from './config.actions';
