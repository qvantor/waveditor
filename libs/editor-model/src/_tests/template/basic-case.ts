import { EditorSnapshot } from '../../types';
import { ConfigFont } from '../../config';
import { Element } from '../../elements';
import { Variable } from '../../variables';

export const MooliFont: ConfigFont = {
  id: '0.1332247839013243',
  url: 'https://fonts.googleapis.com/css2?family=Mooli&display=swap',
  name: 'Mooli',
  fallback: 'Helvetica',
  genericFamily: 'serif',
};

export const InterFont: ConfigFont = {
  id: '0.7248257932515505',
  url: 'https://fonts.googleapis.com/css2?family=Inter&display=swap',
  name: 'Inter',
  fallback: 'Helvetica',
  genericFamily: 'sans-serif',
};

export const NameVariable: Variable = {
  id: '0.7830225981794139',
  type: 'string',
  label: 'name',
  defaultValue: 'John Doe',
};

export const AgeVariable: Variable = {
  id: '0.24456524413525726',
  type: 'string',
  label: 'age',
  defaultValue: '32',
};

export const SimpleLayoutImage: Element = {
  id: '0.5826978971156556',
  link: null,
  meta: {
    width: 626,
    height: 626,
  },
  name: 'image1',
  type: 'image',
  style: {},
  params: {
    url: 'https://img.freepik.com/free-vector/choose-concept-illustration_114360-553.jpg?size=626&ext=jpg',
  },
};

export const SimpleLayoutText: Element = {
  id: '0.8946935213422873',
  link: null,
  name: 'text4',
  type: 'text',
  style: {},
  params: {
    content: {
      type: 'doc',
      content: [
        {
          type: 'paragraph',
          content: [
            {
              text: 'Simple layout',
              type: 'text',
            },
          ],
        },
      ],
    },
  },
};

export const ImageId = '0.6057618679481303';
export const SimpleTextId = '0.8946935213422873';
export const TextWithMooliFont = '0.8154554735234014';
export const TextWithNameVariable = '0.47976090123156934';
export const TextWithNameVariableAndMooliFont = '0.5151065661804399';
export const TextWithTwoVariablesAndInterFont = '0.09364994320533238';
export const SimpleLayoutOfThree = '0.6750283325075193';
export const LayoutWith2FontsAnd2Variables = '0.970902515713419';
export const Template: EditorSnapshot = {
  config: {
    fonts: [
      {
        id: '0.24797196662515453',
        fallback: 'Helvetica',
        genericFamily: 'sans-serif',
      },
      MooliFont,
      InterFont,
    ],
    style: {
      margin: '0',
      fontSize: '16px',
      backgroundColor: '#f3f3f5',
    },
    rootElementId: '0.38109618781050725',
    viewportWidth: 600,
  },
  elements: {
    '0.970902515713419': {
      id: '0.970902515713419',
      link: null,
      name: 'layout2',
      type: 'layout',
      style: {
        verticalAlign: 'top',
      },
      params: {
        gap: 20,
        columns: [
          {
            children: ['0.6057618679481303'],
            proportion: 39.74,
          },
          {
            children: ['0.8893417274842137', '0.6487191447249776'],
            proportion: 60.26,
          },
        ],
      },
    },
    '0.992109257616949': {
      id: '0.992109257616949',
      link: null,
      name: 'text',
      type: 'text',
      style: {},
      params: {
        content: {
          type: 'doc',
          content: [
            {
              type: 'paragraph',
              content: [
                {
                  text: 'Empty text',
                  type: 'text',
                },
              ],
            },
          ],
        },
      },
    },
    '0.5151065661804399': {
      id: '0.5151065661804399',
      link: null,
      name: 'text3',
      type: 'text',
      style: {},
      params: {
        content: {
          type: 'doc',
          content: [
            {
              type: 'paragraph',
              content: [
                {
                  text: 'Text with Mooli Font and ',
                  type: 'text',
                },
                {
                  type: 'variable',
                  attrs: {
                    id: '0.7830225981794139',
                    label: 'name',
                  },
                },
                {
                  text: ' ',
                  type: 'text',
                },
              ],
            },
          ],
        },
      },
    },
    '0.5826978971156556': SimpleLayoutImage,
    '0.6057618679481303': {
      id: '0.6057618679481303',
      link: null,
      meta: {
        width: 626,
        height: 626,
      },
      name: 'image2',
      type: 'image',
      style: {},
      params: {
        url: 'https://img.freepik.com/free-vector/grocery-pick-up-service-abstract-concept-vector-illustration-online-grocery-ordering-virus-protected-shopping-fresh-safe-products-express-food-delivery-ecommerce-abstract-metaphor_335657-2932.jpg?size=626&ext=jpg',
      },
    },
    '0.6487191447249776': {
      id: '0.6487191447249776',
      link: null,
      name: 'layout3',
      type: 'layout',
      style: {
        height: '170px',
        verticalAlign: 'bottom',
      },
      params: {
        columns: [
          {
            children: ['0.7402764635803678'],
            proportion: 50,
          },
          {
            children: ['0.09364994320533238'],
            proportion: 50,
          },
        ],
      },
    },
    '0.6750283325075193': {
      id: '0.6750283325075193',
      link: null,
      name: 'layout1',
      type: 'layout',
      style: {},
      params: {
        columns: [
          {
            children: ['0.5826978971156556'],
            proportion: 50,
          },
          {
            children: ['0.8946935213422873'],
            proportion: 50,
          },
        ],
      },
    },
    '0.7402764635803678': {
      id: '0.7402764635803678',
      link: null,
      name: 'text6',
      type: 'text',
      style: {},
      params: {
        content: {
          type: 'doc',
          content: [
            {
              type: 'paragraph',
              content: [
                {
                  text: 'Fonts is here',
                  type: 'text',
                },
              ],
            },
          ],
        },
      },
    },
    '0.7517803589941807': {
      id: '0.7517803589941807',
      link: null,
      meta: {
        width: 626,
        height: 417,
      },
      name: 'image',
      type: 'image',
      style: {
        height: '150px',
      },
      params: {
        url: 'https://img.freepik.com/free-vector/pick-up-truck-concept-illustration_114360-7813.jpg?size=626&ext=jpg',
      },
    },
    '0.8154554735234014': {
      id: '0.8154554735234014',
      link: null,
      name: 'text2',
      type: 'text',
      style: {},
      params: {
        content: {
          type: 'doc',
          content: [
            {
              type: 'paragraph',
              content: [
                {
                  text: 'Text with ',
                  type: 'text',
                },
                {
                  text: 'Mooli',
                  type: 'text',
                  marks: [
                    {
                      type: 'underline',
                    },
                  ],
                },
                {
                  text: ' font',
                  type: 'text',
                },
              ],
            },
          ],
        },
      },
    },
    '0.8893417274842137': {
      id: '0.8893417274842137',
      link: null,
      name: 'text5',
      type: 'text',
      style: {},
      params: {
        content: {
          type: 'doc',
          content: [
            {
              type: 'paragraph',
              content: [
                {
                  text: 'Layout with variables and fonts',
                  type: 'text',
                },
              ],
            },
          ],
        },
      },
    },
    '0.8946935213422873': SimpleLayoutText,
    '0.09364994320533238': {
      id: '0.09364994320533238',
      link: null,
      name: 'text7',
      type: 'text',
      style: {},
      params: {
        content: {
          type: 'doc',
          content: [
            {
              type: 'paragraph',
              content: [
                {
                  text: 'A have ',
                  type: 'text',
                },
                {
                  type: 'variable',
                  attrs: {
                    id: '0.7830225981794139',
                    label: 'name',
                  },
                },
                {
                  text: ' and ',
                  type: 'text',
                },
                {
                  type: 'variable',
                  attrs: {
                    id: '0.24456524413525726',
                    label: 'age',
                  },
                },
                {
                  text: ' and ',
                  type: 'text',
                },
                {
                  text: 'Inter',
                  type: 'text',
                  marks: [
                    {
                      type: 'italic',
                    },
                  ],
                },
                {
                  text: ' font',
                  type: 'text',
                },
              ],
            },
          ],
        },
      },
    },
    '0.38109618781050725': {
      id: '0.38109618781050725',
      link: null,
      name: 'layout',
      type: 'layout',
      style: {
        margin: '0px auto',
        padding: '20px',
        backgroundColor: '#fff',
      },
      params: {
        columns: [
          {
            children: [
              '0.992109257616949',
              '0.7517803589941807',
              '0.8154554735234014',
              '0.47976090123156934',
              '0.5151065661804399',
              '0.6750283325075193',
              '0.970902515713419',
            ],
            proportion: 100,
          },
        ],
      },
    },
    '0.47976090123156934': {
      id: '0.47976090123156934',
      link: null,
      name: 'text1',
      type: 'text',
      style: {},
      params: {
        content: {
          type: 'doc',
          content: [
            {
              type: 'paragraph',
              content: [
                {
                  text: 'hello ',
                  type: 'text',
                },
                {
                  type: 'variable',
                  attrs: {
                    id: '0.7830225981794139',
                    label: 'name',
                  },
                },
                {
                  text: ' ',
                  type: 'text',
                },
              ],
            },
          ],
        },
      },
    },
  },
  relations: {
    elementFont: {
      '0.5151065661804399': '0.1332247839013243',
      '0.7402764635803678': '0.1332247839013243',
      '0.8154554735234014': '0.1332247839013243',
      '0.09364994320533238': '0.7248257932515505',
    },
  },
  variables: [
    NameVariable,
    AgeVariable,
    { id: '0', label: 'Unused Variable', type: 'string' },
  ],
};
