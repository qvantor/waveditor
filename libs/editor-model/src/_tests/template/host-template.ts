import { EditorSnapshot } from '../../types';
import { ConfigFont } from '../../config';

export const MooliFont: ConfigFont = {
  id: '0.43368360441087606',
  url: 'https://fonts.googleapis.com/css2?family=Mooli&display=swap',
  name: 'Mooli',
  fallback: 'Helvetica',
  genericFamily: 'sans-serif',
};
export const Template: EditorSnapshot = {
  config: {
    fonts: [
      MooliFont,
      {
        id: '0.4317102971377873',
        fallback: 'Times New Roman',
        genericFamily: 'serif',
      },
    ],
    style: {
      margin: '0',
      fontSize: '16px',
      backgroundColor: '#f3f3f5',
    },
    rootElementId: '0.4464062060314937',
    viewportWidth: 600,
  },
  elements: {
    '0.4464062060314937': {
      id: '0.4464062060314937',
      link: null,
      name: 'layout',
      type: 'layout',
      style: {
        margin: '0px auto',
        padding: '20px',
        backgroundColor: '#fff',
      },
      params: {
        columns: [
          {
            children: ['0.10827230063110993', '0.5788179525717478'],
            proportion: 100,
          },
        ],
      },
    },
    '0.5788179525717478': {
      id: '0.5788179525717478',
      link: null,
      name: 'text',
      type: 'text',
      style: {},
      params: {
        content: {
          type: 'doc',
          content: [
            {
              type: 'paragraph',
              content: [
                {
                  text: 'Some text written with mooli',
                  type: 'text',
                },
              ],
            },
          ],
        },
      },
    },
    '0.10827230063110993': {
      id: '0.10827230063110993',
      link: null,
      name: 'text1',
      type: 'text',
      style: {},
      params: {
        content: {
          type: 'doc',
          content: [
            {
              type: 'paragraph',
              content: [
                {
                  text: 'Some text with ',
                  type: 'text',
                },
                {
                  type: 'variable',
                  attrs: {
                    id: '0.6910150707471912',
                    label: 'name',
                  },
                },
                {
                  text: ' variable',
                  type: 'text',
                },
              ],
            },
          ],
        },
      },
    },
  },
  relations: {
    elementFont: {
      '0.10827230063110993': '0.43368360441087606',
    },
  },
  variables: [
    {
      id: '0.6910150707471912',
      type: 'string',
      label: 'name',
      defaultValue: 'Some name',
    },
  ],
};
