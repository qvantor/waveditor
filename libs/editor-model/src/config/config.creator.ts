import { generateId } from '@waveditors/utils';
import { Config, ConfigFont } from './config.types';

export const createInitialConfig = (rootElementId: string): Config => {
  return {
    rootElementId,
    viewportWidth: 600,
    style: {
      margin: '0',
      backgroundColor: '#f3f3f5',
      fontSize: '16px',
    },
    fonts: [
      {
        id: generateId(),
        fallback: 'Helvetica',
        genericFamily: 'sans-serif',
      },
      {
        id: generateId(),
        fallback: 'Times New Roman',
        genericFamily: 'serif',
      },
    ],
  };
};

export const createConfigFont = (font: Partial<ConfigFont>): ConfigFont => ({
  id: generateId(),
  fallback: 'Helvetica',
  genericFamily: 'sans-serif',
  ...font,
});
