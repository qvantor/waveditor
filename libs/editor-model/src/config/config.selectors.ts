import { Config } from './config.types';

export const getConfigFonts = (config: Config) => config.fonts;

export const getConfigFontById = (id: string) => (config: Config) =>
  getConfigFonts(config).find((font) => font.id === id);

export const getTemplateDefaultFont = (config: Config) => {
  const defaultFont = getConfigFonts(config)[0];
  if (!defaultFont) throw new Error('Model corrupted: no default font found');
  return defaultFont;
};

export const getConfigName = (config: Config) => config.name;
export const getConfigRootElementId = (config: Config) => config.rootElementId;

export const getConfigViewportWidth = (config: Config) => config.viewportWidth;
