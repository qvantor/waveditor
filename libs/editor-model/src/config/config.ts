import { createStore, StoreResult } from '@waveditors/rxjs-react';
import { Config, ConfigFont, FontChangedPayload } from './config.types';

export const configStoreConstructor = () =>
  createStore<Config>().addActions({
    setName: (name: string, state) => ({ ...state, name }),
    addFont: (font: ConfigFont, state) => ({
      ...state,
      fonts: [...state.fonts, font],
    }),
    removeFont: (fontId: string, state) => ({
      ...state,
      fonts: state.fonts.filter((font) => font.id !== fontId),
    }),
    setFont: ({ id, value }: FontChangedPayload, state) => ({
      ...state,
      fonts: state.fonts.map((font) => {
        if (font.id !== id) return font;
        return value;
      }),
    }),
  });

export type ConfigStore = StoreResult<typeof configStoreConstructor>;
