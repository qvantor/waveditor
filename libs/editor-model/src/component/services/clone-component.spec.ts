import { mapJSONContent } from '@waveditors/utils';
import { ConfigFont } from '../../config';
import { BasicCase } from '../../_tests';
import { TextWithTwoVariablesAndInterFont } from '../../_tests/template/basic-case';
import {
  cloneComponent,
  applyFontsIdTableToRelations,
  applyVariablesTableToElements,
} from './clone-component';

const { Template, InterFont, MooliFont, NameVariable } = BasicCase;
describe('clone component', () => {
  it('should not affect counts', () => {
    const clone = cloneComponent(Template);
    expect(Object.keys(clone).length).toBe(Object.keys(Template).length);
    expect(Object.keys(clone.relations.elementFont).length).toBe(
      Object.keys(Template.relations.elementFont).length
    );
    expect(clone.variables.length).toBe(Template.variables.length);
    expect(clone.config.fonts.length).toBe(Template.config.fonts.length);
  });
  it('should create new elements ids', () => {
    const clone = cloneComponent(Template);
    Object.keys(Template.elements).forEach((id) =>
      expect(clone.elements[id]).toBeUndefined()
    );
  });
  it('should create new fonts ids', () => {
    const clone = cloneComponent(Template);
    expect(clone.config.fonts).toEqual(
      expect.arrayContaining(
        Template.config.fonts.map(({ id, ...font }) =>
          expect.objectContaining(font)
        )
      )
    );
  });
  it('should create new variables ids', () => {
    const clone = cloneComponent(Template);
    expect(clone.variables).toEqual(
      expect.arrayContaining(
        Template.variables.map(({ id, ...font }) =>
          expect.objectContaining(font)
        )
      )
    );
  });
  it('should keep font-element relations', () => {
    const clone = cloneComponent(Template);
    const mooliFonts = ['text3', 'text6', 'text2'];
    const interFont = 'text7';

    const { id: MooliFontNewId } = clone.config.fonts.find(
      (font) => font.name === MooliFont.name
    ) as ConfigFont;
    const { id: InterFontNewId } = clone.config.fonts.find(
      (font) => font.name === InterFont.name
    ) as ConfigFont;

    const interRelations = Object.entries(clone.elements)
      .filter(([_, element]) => element.name === interFont)
      .map(([id]) => id)
      .reduce((sum, id) => ({ ...sum, [id]: InterFontNewId }), {});

    const mooliRelations = Object.entries(clone.elements)
      .filter(([_, element]) => mooliFonts.includes(element.name as string))
      .map(([id]) => id)
      .reduce((sum, id) => ({ ...sum, [id]: MooliFontNewId }), {});

    expect(clone.relations.elementFont).toEqual({
      ...mooliRelations,
      ...interRelations,
    });
  });
  it('should update all variables in elements', () => {
    const clone = cloneComponent(Template);
    const varIds = clone.variables.map((variable) => variable.id);
    Object.values(clone.elements).forEach((element) => {
      if (element.type !== 'text') return;
      mapJSONContent(element.params.content, (item) => {
        if (item.type !== 'variable') return item;
        expect(varIds).toContain(item.attrs?.id);
        return item;
      });
    });
  });
  it('should update all layouts children', () => {
    const clone = cloneComponent(Template);
    Object.values(clone.elements).forEach((element) => {
      if (element.type !== 'layout') return;
      element.params.columns.forEach((col) => {
        col.children.forEach((childId) =>
          expect(clone.elements[childId]).toBeDefined()
        );
      });
    });
  });

  it('applyFontsIdTableToRelations should work with partial table', () => {
    const clonedRelations = applyFontsIdTableToRelations(Template.relations, {
      [MooliFont.id]: 'test',
    });
    expect(clonedRelations).toMatchSnapshot();
  });
  it('applyVariablesTableToElements should work with partial table', () => {
    const clonedElements = applyVariablesTableToElements(Template.elements, {
      [NameVariable.id]: 'test',
    });
    expect(
      clonedElements[TextWithTwoVariablesAndInterFont].params
    ).toMatchSnapshot();
  });
});
