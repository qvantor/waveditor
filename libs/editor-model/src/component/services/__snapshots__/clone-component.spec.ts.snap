// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clone component applyFontsIdTableToRelations should work with partial table 1`] = `
Object {
  "elementFont": Object {
    "0.09364994320533238": "0.7248257932515505",
    "0.5151065661804399": "test",
    "0.7402764635803678": "test",
    "0.8154554735234014": "test",
  },
}
`;

exports[`clone component applyVariablesTableToElements should work with partial table 1`] = `
Object {
  "content": Object {
    "content": Array [
      Object {
        "content": Array [
          Object {
            "text": "A have ",
            "type": "text",
          },
          Object {
            "attrs": Object {
              "id": "test",
              "label": "name",
            },
            "type": "variable",
          },
          Object {
            "text": " and ",
            "type": "text",
          },
          Object {
            "attrs": Object {
              "id": "0.24456524413525726",
              "label": "age",
            },
            "type": "variable",
          },
          Object {
            "text": " and ",
            "type": "text",
          },
          Object {
            "marks": Array [
              Object {
                "type": "italic",
              },
            ],
            "text": "Inter",
            "type": "text",
          },
          Object {
            "text": " font",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
    ],
    "type": "doc",
  },
}
`;
